Add options to use external libraries of nlohmann-json and cxxopts

topazus opened the pull request:
In this PR, 
- Add options `F3D_USE_EXTERNAL_CXXOPTS` and `F3D_USE_EXTERNAL_NLOHMANN_JSON` (both are off by default) for third party libraries of nlohmann_json and cxxopts to choose their correspoding system libraries. 
- Move the json.hpp file to `external/nlohmann` directory to unify the include directive, cxxopts.hpp file to `external/cxxopts` directory.
- Updated to cxxopts to 3.1.1 version. The cxxopts package is 3.1.1 version in Fedora repository, so I updated it to this version accordingly, ref: https://packages.fedoraproject.org/pkgs/cxxopts/cxxopts-devel/ . Previously in cxxopts, the exceptions derives from cxxopts::OptionException, but now for 3.1.1 it derives from cxxopts::exceptions::exception, ref: https://github.com/jarro2783/cxxopts/pull/347

try to resolve: #830

Meakk commented on CMakeLists.txt:
```diff
@@ -205,6 +205,8 @@ if(EMSCRIPTEN)
   add_subdirectory(webassembly)
 endif ()
 
+option(F3D_USE_EXTERNAL_DEPS "Use external dependencies" OFF)
```
Would be more convenient if we had two different options `F3D_USE_EXTERNAL_CXXOPTS` and `F3D_USE_EXTERNAL_NLOHMANN_JSON`.

topazus replied:
Thank for the suggestion, I'll use these two options later.

Meakk commented on application/CMakeLists.txt:
```diff
@@ -56,11 +56,20 @@ elseif (MSVC)
+if (F3D_USE_EXTERNAL_DEPS)
+  find_package(cxxopts REQUIRED)
+  find_package(nlohmann_json REQUIRED)
+  target_include_directories(f3d
```
I think you need `target_link_libraries(f3d PRIVATE cxxopts::cxxopts)` and `target_link_libraries(f3d PRIVATE nlohmann_json::nlohmann_json)` to add the include directories of these dependencies in case the install directory is not standard.  
I'm not sure about the target names, so please double check.

mwestphal replied:
arent these libs headeronly ?

Meakk replied:
Indeed, but the targets expose the include directories via `INCLUDE_DIRECTORIES` target property set up by the target config file

Meakk commented:
@mwestphal I vaguely remember you had modified cxxopts. Do you remember what you did? I just want to be sure the update is not breaking anything in F3D.

Meakk commented:
@topazus you have to update `.cppcheck.supp` with the new path to json.hpp

topazus commented:
> @mwestphal I vaguely remember you had modified cxxopts. Do you remember what you did? I just want to be sure the update is not breaking anything in F3D.

Because the cxxopts package is 3.1.1 version in Fedora repository, so I updated it to this version accordingly, ref: https://packages.fedoraproject.org/pkgs/cxxopts/cxxopts-devel/
I modified the catch exception of the cxxopts usage. Previously in cxxopts, the exceptions derives from `cxxopts::OptionException`, but now for 3.1.1 it derives from `cxxopts::exceptions::exception`, ref: https://github.com/jarro2783/cxxopts/pull/347

mwestphal commented on external/cxxopts.hpp:
```diff
@@ -1,6 +1,6 @@
 /*
 
-Copyright (c) 2014, 2015, 2016, 2017 Jarryd Beck
+Copyright (c) 2014-2022 Jarryd Beck
```
@Meakk About cxxopts pathch, here is what I found.

1. There is currently no patch in cxxopts, so this PR should move forward withoout issues
2. I did had some troubles two years ago when reoworking the option system and opened some issues on cxxopts repo, but they were ultimaly closed as it was misuse on my part: https://github.com/jarro2783/cxxopts/issues?q=is%3Aissue+author%3Amwestphal+is%3Aclosed
3. At the time, there was a patch, but it may have been a local patch only and was never merged
4. Options rework is visible here, notic the assert to ensure we do not make the same mistake: https://github.com/f3d-app/f3d/pull/396/files#diff-7c425fb94271a2540a78bd3b67b26cf5f90dd107db21df60aaf3a57137b379f2
5. Some time later, you removed the assert and I approved it, because it was catching to wide a net and in some case this is supported.

The only unsupported case is vector<string> empty initialization, that we are currently avoiding.

So all good !

Meakk replied:
Good

mwestphal commented on application/F3DOptionsParser.cxx:
```diff
@@ -6,7 +6,7 @@
 
 #include "cxxopts.hpp"
-#include "json.hpp"
+#include "nlohmann/json.hpp"
```
why moving one and not the other ?

topazus replied:
> why moving one and not the other ?

their header files are at different directories.

```bash
[ruby@fedora fedora-src]$ rpm -ql cxxopts-devel | grep /usr/include/cxxopts.hpp
/usr/include/cxxopts.hpp
[ruby@fedora fedora-src]$ rpm -ql json-devel | grep /usr/include/nlohmann/json.hpp
/usr/include/nlohmann/json.hpp
```

mwestphal replied:
Ok, thats fair. Seems also the case on ArchLinux.

mwestphal commented on application/CMakeLists.txt:
```diff
@@ -56,11 +56,27 @@ elseif (MSVC)
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
   )
+else ()
+  target_include_directories(f3d
```
you can do multiple target_include_directories calls I think

topazus replied:
I do not get your point here. Could you say more expliciltly?

mwestphal replied:
you could write it like this, if you put the headers in their own dir

```
target_include_directories(f3d PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
if (F3D_USE_EXTERNAL_CXXOPTS)
  find_package(cxxopts REQUIRED)
  target_link_libraries(f3d PRIVATE cxxopts::cxxopts)
else ()
  target_include_directories(f3d PUBLIC $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/external/cxxopts>
endif ()

if (F3D_USE_EXTERNAL_NLOHMANN_JSON)
  find_package(nlohmann_json REQUIRED)
  target_link_libraries(f3d PRIVATE nlohmann_json::nlohmann_json)
else ()
  target_include_directories(f3d PUBLIC $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/external/json>
endif ()
```

This way we would be sure of which include is done, which is not the case currently

topazus commented:
________________________________
From: Mathieu Westphal ***@***.***>
Sent: Sunday, May 28, 2023 11:02:28 PM
To: f3d-app/f3d ***@***.***>
Cc: Felix Wang ***@***.***>; Mention ***@***.***>
Subject: Re: [f3d-app/f3d] Add options to use external libraries of nlohmann-json and cxxopts (PR #832)


@mwestphal commented on this pull request.

________________________________

In application/CMakeLists.txt<https://github.com/f3d-app/f3d/pull/832#discussion_r1208572665>:

>    )
+else ()
+  target_include_directories(f3d


you can do multiple target_include_directories calls I think

—
Reply to this email directly, view it on GitHub<https://github.com/f3d-app/f3d/pull/832#pullrequestreview-1448287583>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASNPISN6XOAPUSDQK2CT4KDXINSIJANCNFSM6AAAAAAYRWQSMQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Meakk reviewed (approved):
LGTM. @mwestphal I let you review and merge when you think it's ready.

mwestphal commented on application/CMakeLists.txt:
```diff
@@ -56,11 +56,20 @@ elseif (MSVC)
+  find_package(nlohmann_json REQUIRED)
+  target_link_libraries(f3d PRIVATE nlohmann_json::nlohmann_json)
+else ()
+  target_include_directories(f3d PUBLIC $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/external>)
```
I dont like that because if we add another external, we may include the wrond header. Please put `nlohmann` dir in a dedicated directory.

mwestphal replied:
https://github.com/f3d-app/f3d/actions/runs/5120469346/jobs/9207007100?pr=832

topazus replied:
I am a little confused about the problem of including the header file.

mwestphal replied:
looking at the code it should work, So I do not know. Are you able to reproduce locally ?

Meakk replied:
nlohmann_json is used in libf3d, not F3D app, see here: https://github.com/f3d-app/f3d/blob/614264de76967461cca02fed3daccb3548b8c5c7/library/CMakeLists.txt#L102

Meakk replied:
Actually, it's used in both. So you also have to change it in libf3d but keep the changes in F3D app.

topazus replied:
Thanks for pointing out the problem.

Meakk commented on CMakeLists.txt:
```diff
@@ -205,6 +205,9 @@ if(EMSCRIPTEN)
   add_subdirectory(webassembly)
 endif ()
 
+option(F3D_USE_EXTERNAL_CXXOPTS "Use external cxxopts dependencies" OFF)
+option(F3D_USE_EXTERNAL_NLOHMANN_JSON "Use external nlohmann_json dependencies" OFF)
```
```suggestion
option(F3D_USE_EXTERNAL_CXXOPTS "Use external cxxopts dependency" OFF)
option(F3D_USE_EXTERNAL_NLOHMANN_JSON "Use external nlohmann_json dependency" OFF)
```

Meakk commented on winshellext/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@ add_library(F3DShellExtension SHARED
 target_compile_definitions(F3DShellExtension PUBLIC UNICODE F3D_WINDOWS_THUMBNAIL_TIMEOUT=8000)
 
-target_include_directories(F3DShellExtension PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_SOURCE_DIR}/external)
+target_include_directories(F3DShellExtension PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_SOURCE_DIR}/external/nlohmann_json)
```
and here as well:

```cmake
if (F3D_USE_EXTERNAL_NLOHMANN_JSON)
  find_package(nlohmann_json REQUIRED)
  target_link_libraries(F3DShellExtension PRIVATE nlohmann_json::nlohmann_json)
else ()
  target_include_directories(F3DShellExtension PRIVATE $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/external/nlohmann_json>)
endif ()
```

topazus replied:
Many thanks.

Meakk reviewed (changes requested):
Sorry I have more comments :)  
It's almost done though!

mwestphal commented on CMakeLists.txt:
```diff
@@ -205,6 +205,13 @@ if(EMSCRIPTEN)
+
+if (F3D_USE_EXTERNAL_NLOHMANN_JSON)
+  find_package(nlohmann_json REQUIRED)
+endif ()
```
lets do it for both

topazus replied:
Done.

Meakk commented on .cppcheck.supp:
```diff
@@ -3,8 +3,8 @@ missingInclude
-*:external/json.hpp
-*:external/cxxopts.hpp
+*:external/nlohmann_json/nlohmann/json.hpp
+*:external/cxxopts/cxxopts.hpp
```
Unrelated to this file, `.codespellrc` must be updated.

mwestphal replied:
right :facepalm:

mwestphal commented:
@topazus looking good ! Lets wait for @Meakk review.

BTW we have a chat dedicated for packager on our discord, do not hesitate to join !

https://discord.f3d.app

Meakk commented:
`find_package(cxxopts REQUIRED)` is called AFTER `add_subdirectory(application)`.  
Move it before, it should work.

topazus commented:
> `find_package(cxxopts REQUIRED)` is called AFTER `add_subdirectory(application)`. Move it before, it should work.

Ah, it was error-free when find_package was placed before `add_subdirectory(application)`. I have another confusion, I only called `find_package(cxxopts REQUIRED)` again in application/CMakeLists.txt file, it was also built successfully, why does no error, which is the target not found, happened for nlohmann_json?

Meakk commented:
nlohmann_json is not used by the application, it is used in the library and the shellext though so the find_package must be called before adding library and shellext subdirectories.

topazus commented:
https://github.com/f3d-app/f3d/blob/57dec2d8ac7cc76cd1fca38eae28c37d6b943721/application/F3DOptionsParser.cxx#L633-L637
`application/F3DOptionsParser.cxx` file includes the nlohmann/json.hpp header file, does this not mean nlohmann_json was used in application subdirectory?

Meakk commented:
You're right. Are you sure it's working if the find_package is called after the add_subdirectory?

topazus commented:
with the following modification:
```bash
ruby in 🌐 fedora in f3d on  master [!?] via △ v3.26.4 
❯ git diff
diff --git a/application/CMakeLists.txt b/application/CMakeLists.txt
index 05e1a08..81781c7 100644
--- a/application/CMakeLists.txt
+++ b/application/CMakeLists.txt
@@ -58,6 +58,7 @@ endif ()
 
 target_include_directories(f3d PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
 if (F3D_USE_EXTERNAL_CXXOPTS)
+  find_package(cxxopts REQUIRED)
   target_link_libraries(f3d PRIVATE cxxopts::cxxopts)
 else ()
   target_include_directories(f3d PUBLIC $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/external/cxxopts>)

ruby in 🌐 fedora in f3d on  master [!?] via △ v3.26.4 
❯ git log
commit 57dec2d8ac7cc76cd1fca38eae28c37d6b943721 (HEAD -> master, origin/master, origin/HEAD)
Author: Felix Wang <77263945+topazus@users.noreply.github.com>
Date:   Thu Jun 1 13:24:43 2023 +0800

    Add options to use external libraries of nlohmann-json and cxxopts (#832)
    
    * Add F3D_USE_EXTERNAL_DEPS option for third party libraries
    
    * Use options F3D_USE_EXTERNAL_CXXOPTS and F3D_USE_EXTERNAL_NLOHMANN_JSON instead
    
    * fix cppcheck
    
    * explicitly specify the include directories
    
    * fix
    
    * fix typo
    
    * add find_package to toplevel CMakeLists.txt file
```

<details>

<summary>build process</summary>

```bash
ruby in 🌐 fedora in f3d on  master [!?] via △ v3.26.4 
❯ rm -rf build/ && cmake . -B build -DF3D_USE_EXTERNAL_CXXOPTS=ON -DF3D_USE_EXTERNAL_NLOHMANN_JSON=ON
-- The C compiler identification is GNU 13.1.1
-- The CXX compiler identification is GNU 13.1.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/lib64/ccache/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/lib64/ccache/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found GLEW: /usr/lib64/libGLEW.so  
-- Found OpenGL: /usr/lib64/libOpenGL.so  found components: OpenGL GLX 
-- Found HDF5: /usr/lib64/libhdf5.so (found version "1.12.1") found components: C HL 
-- Found utf8cpp: /usr/include/utf8cpp  
-- Found NetCDF: netcdf (found version "4.9.0") 
-- Found JsonCpp: /usr/lib64/libjsoncpp.so (found suitable version "1.9.5", minimum required is "0.7.0") 
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.13")  
-- Found PNG: /usr/lib64/libpng.so (found version "1.6.37") 
-- Found nlohmann_json: /usr/share/cmake/nlohmann_json/nlohmann_jsonConfig.cmake (found version "3.11.2") 
-- Found Eigen3: /usr/include/eigen3 (found version "3.4.0") 
-- Found X11: /usr/include   
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found EXPAT: /usr/lib64/libexpat.so (found version "2.5.0") 
-- Found double-conversion: /usr/lib64/libdouble-conversion.so  
-- Found LZ4: /usr/lib64/liblz4.so (found version "1.9.4") 
-- Found LZMA: /usr/lib64/liblzma.so (found version "5.4.3") 
-- Found JPEG: /usr/lib64/libjpeg.so (found version "62") 
-- Found TIFF: /usr/lib64/libtiff.so (found version "4.4.0")  
-- Found Freetype: /usr/lib64/libfreetype.so (found version "2.13.0") 
-- VTK 9.2.6 found
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
-- Configuring done (2.6s)
-- Generating done (0.0s)
-- Build files have been written to: /home/ruby/git-repo/f3d/build

ruby in 🌐 fedora in f3d on  master [!?] via △ v3.26.4 took 2s 
❯ cmake --build build/ -j6
[  7%] Building CXX object library/VTKExtensions/Core/CMakeFiles/VTKExtensionsCore.dir/F3DLog.cxx.o
[  9%] Building CXX object plugins/native/CMakeFiles/f3d-plugin-native.dir/plugin.cxx.o
[ 11%] Generating the wrap hierarchy for f3d::VTKExtensionsReaders
[  9%] Generating the wrap hierarchy for f3d::VTKExtensionsApplicative
[ 11%] Generating the wrap hierarchy for f3d::VTKExtensionsCore
[ 11%] Building CXX object plugins/exodus/CMakeFiles/f3d-plugin-exodus.dir/plugin.cxx.o
[ 15%] Linking CXX static library ../../lib/libf3d-plugin-exodus.a
[ 15%] Building CXX object library/VTKExtensions/Core/CMakeFiles/VTKExtensionsCore.dir/vtkF3DConsoleOutputWindow.cxx.o
[ 15%] Built target VTKExtensionsCore-hierarchy
[ 17%] Linking CXX static library ../../lib/libf3d-plugin-native.a
[ 17%] Built target VTKExtensionsApplicative-hierarchy
[ 19%] Linking CXX static library ../../../lib/libVTKExtensionsCore.a
[ 19%] Built target VTKExtensionsReaders-hierarchy
[ 21%] Generating the wrap hierarchy for f3d::VTKExtensionsRendering
[ 21%] Built target f3d-plugin-exodus
[ 21%] Built target f3d-plugin-native
[ 21%] Built target VTKExtensionsCore
[ 25%] Building CXX object library/VTKExtensions/Readers/CMakeFiles/VTKExtensionsReaders.dir/vtkF3DPostProcessFilter.cxx.o
[ 25%] Building CXX object library/VTKExtensions/Readers/CMakeFiles/VTKExtensionsReaders.dir/vtkF3DGenericImporter.cxx.o
[ 26%] Linking CXX static library ../../../lib/libVTKExtensionsReaders.a
[ 26%] Built target VTKExtensionsReaders
[ 32%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DCachedLUTTexture.cxx.o
[ 32%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DCachedSpecularTexture.cxx.o
[ 36%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DInteractorEventRecorder.cxx.o
[ 32%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DDropZoneActor.cxx.o
[ 34%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DHexagonalBokehBlurPass.cxx.o
[ 42%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DInteractorStyle.cxx.o
[ 42%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DOpenGLGridMapper.cxx.o
[ 44%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DPolyDataMapper.cxx.o
[ 42%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DNoRenderWindow.cxx.o
[ 46%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DRenderPass.cxx.o
[ 46%] Built target VTKExtensionsRendering-hierarchy
[ 48%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DRenderer.cxx.o
[ 51%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DOrientationMarkerWidget.cxx.o
[ 51%] Building CXX object library/VTKExtensions/Rendering/CMakeFiles/VTKExtensionsRendering.dir/vtkF3DRendererWithColoring.cxx.o
[ 53%] Linking CXX static library ../../../lib/libVTKExtensionsRendering.a
[ 53%] Built target VTKExtensionsRendering
[ 55%] Building CXX object library/VTKExtensions/Applicative/CMakeFiles/VTKExtensionsApplicative.dir/vtkF3DObjectFactory.cxx.o
[ 57%] Linking CXX static library ../../../lib/libVTKExtensionsApplicative.a
[ 57%] Built target VTKExtensionsApplicative
[ 65%] Building CXX object library/CMakeFiles/libf3d.dir/src/camera_impl.cxx.o
[ 65%] Building CXX object library/CMakeFiles/libf3d.dir/src/engine.cxx.o
[ 63%] Building CXX object library/CMakeFiles/libf3d.dir/src/animationManager.cxx.o
[ 65%] Building CXX object library/CMakeFiles/libf3d.dir/src/config.cxx.o
[ 67%] Building CXX object library/CMakeFiles/libf3d.dir/src/factory.cxx.o
[ 69%] Building CXX object library/CMakeFiles/libf3d.dir/src/image.cxx.o
[ 78%] Building CXX object library/CMakeFiles/libf3d.dir/src/interactor_impl.cxx.o
[ 76%] Building CXX object library/CMakeFiles/libf3d.dir/src/init.cxx.o
[ 78%] Building CXX object library/CMakeFiles/libf3d.dir/src/loader_impl.cxx.o
[ 78%] Building CXX object library/CMakeFiles/libf3d.dir/src/log.cxx.o
[ 78%] Building CXX object library/CMakeFiles/libf3d.dir/src/interactor.cxx.o
[ 82%] Building CXX object library/CMakeFiles/libf3d.dir/src/window_impl.cxx.o
[ 82%] Building CXX object library/CMakeFiles/libf3d.dir/src/options.cxx.o
[ 84%] Linking CXX shared library ../lib/libf3d.so
[ 84%] Built target libf3d
[ 86%] Generating F3DIcon.h, F3DIcon.cxx
[ 92%] Building CXX object application/CMakeFiles/f3d.dir/F3DOptionsParser.cxx.o
[ 94%] Building CXX object application/CMakeFiles/f3d.dir/F3DStarter.cxx.o
[ 92%] Building CXX object application/CMakeFiles/f3d.dir/F3DIcon.cxx.o
[ 96%] Building CXX object application/CMakeFiles/f3d.dir/F3DSystemTools.cxx.o
[ 92%] Building CXX object application/CMakeFiles/f3d.dir/F3DConfigFileTools.cxx.o
[ 98%] Building CXX object application/CMakeFiles/f3d.dir/main.cxx.o
[100%] Linking CXX executable ../bin/f3d
[100%] Built target f3d
```

</details>

Meakk commented:
That's weird. I'll take a closer look.

Meakk commented:
I think that's because VTK is using `nlohmann_json` and indirectly define the target.  
If you look at `/usr/lib/cmake/vtk/VTK-targets.cmake` (or wherever VTK is installed), you'll find:

```cmake
# Create imported target VTK::nlohmannjson
add_library(VTK::nlohmannjson INTERFACE IMPORTED)

set_target_properties(VTK::nlohmannjson PROPERTIES
  INTERFACE_LINK_LIBRARIES "nlohmann_json::nlohmann_json"
)
```

mwestphal commented:
https://github.com/f3d-app/f3d/pull/840

topazus commented:
@Meakk Your last comment is very helpful to clarify the confusion, thanks. I also see the similar thing at `/usr/lib64/cmake/vtk/VTK-targets.cmake` file of my installed vtk-devel package.
